src/types/types.zig

```diff
@@ -0,0 +1,376 @@
+/// that ends up being inefficient because every number type requires
+/// multiple different type entries. By special casing them here we can
+/// ensure that they have more compact representations
+pub const Num = union(enum) {
+    const Self = @This();
+
+    flex_var,
+    int: Int,
+    frac: Frac,
+
+    /// the Frac data type
+    pub const Frac = union(enum) {
+        flex_var,
+        exact: Precision,
+
+        /// the precision of a frac
+        pub const Precision = enum { f32, f64, dec };
+    };
+
+    /// the Int data type
+    pub const Int = union(enum) {
+        flex_var,
+        exact: Precision,
+
+        /// the precision of an int
+        pub const Precision = enum { u8, i8, u16, i16, u32, i32, u64, i64, u128, i128 };
+    };
+};
+
+/// a num flex_var
+pub const num_flex_var: FlatType = .{ .num = Num.flex_var };
+
+/// a frac flex_var
+pub const frac_flex_var: FlatType = .{ .num = Num{ .frac = .flex_var } };
+
+/// a frac f32
+pub const frac_f32: FlatType = .{ .num = Num{ .frac = .{ .exact = .f32 } } };
+
+/// a frac f64
+pub const frac_f64: FlatType = .{ .num = Num{ .frac = .{ .exact = .f64 } } };
+
+/// a frac dec
+pub const frac_dec: FlatType = .{ .num = Num{ .frac = .{ .exact = .dec } } };
+
+/// a int flex_var
+pub const int_flex_var: FlatType = .{ .num = Num{ .int = .flex_var } };
+
+/// an int u8
+pub const int_u8: FlatType = .{ .num = Num{ .int = .{ .exact = .u8 } } };
+
+/// an int i8
+pub const int_i8: FlatType = .{ .num = Num{ .int = .{ .exact = .i8 } } };
+
+/// an int u16
+pub const int_u16: FlatType = .{ .num = Num{ .int = .{ .exact = .u16 } } };
+
+/// an int i16
+pub const int_i16: FlatType = .{ .num = Num{ .int = .{ .exact = .i16 } } };
+
+/// an int u32
+pub const int_u32: FlatType = .{ .num = Num{ .int = .{ .exact = .u32 } } };
+
+/// an int i32
+pub const int_i32: FlatType = .{ .num = Num{ .int = .{ .exact = .i32 } } };
+
+/// an int u64
+pub const int_u64: FlatType = .{ .num = Num{ .int = .{ .exact = .u64 } } };
+
+/// an int i64
+pub const int_i64: FlatType = .{ .num = Num{ .int = .{ .exact = .i64 } } };
+
+/// an int u128
+pub const int_u128: FlatType = .{ .num = Num{ .int = .{ .exact = .u128 } } };
+
+/// an int i128
+pub const int_i128: FlatType = .{ .num = Num{ .int = .{ .exact = .i128 } } };
+
+// custom types //
+
+/// A nominal user-defined type
+pub const CustomType = struct {
+    ident: TypeIdent,
+    args: Var.SafeList.Range,
+    backing_var: Var,
+};
+
+// functions //
+
+/// Represents a function
+pub const Func = struct {
+    args: Var.SafeList.Range,
+    ret: Var,
+    eff: Var,
+};
+
+// records //
+
+/// Represents a record
+pub const Record = struct {
+    fields: RecordField.SafeMultiList.Range,
+    ext: Var,
+
+    const Self = @This();
+};
+
+/// A field on a record
+pub const RecordField = struct {
+    const Self = @This();
+
+    /// The name of the field
+    name: Ident.Idx,
+    var_: Var,
+
+    /// The context need by sortByFieldNameAsc
```

Comment:
```suggestion
    /// The context needed by sortByFieldNameAsc
```

Reply:
Also, do we need this to be a struct with a single field instead of just `*const Ident.Store`? 🤔